envisage/extension_point.py

```diff
@@ -200,23 +200,41 @@ def connect(self, obj, trait_name):
         """
 
         def listener(extension_registry, event):
-            """ Listener called when an extension point is changed. """
-
-            # If an index was specified then we fire an '_items' changed event.
+            """ Listener called when an extension point is changed.
+
+            Parameters
+            ----------
+            extension_registry : IExtensionRegistry
+                Registry that maintains the extensions.
+            event : ExtensionPointChangedEvent
+                Event for created for the change.
+                If the event.index is None, this means the entire extensions
+                is set to a new value. If the event.index is not None, some
+                portion of the list has been modified.
+            """
             if event.index is not None:
-                name = trait_name + "_items"
-                old = Undefined
-                new = event
+                # We know where in the list is changed.
 
-            # Otherwise, we fire a normal trait changed event.
-            else:
-                name = trait_name
-                old = event.removed
-                new = event.added
+                # Mutate the _ExtensionPointValue to fire ListChangeEvent
+                # expected from observing item change.
+                getattr(obj, trait_name)._sync_values(event)
 
-            obj.trait_property_changed(name, old, new)
+                # For on_trait_change('name_items')
+                obj.trait_property_changed(
+                    trait_name + "_items", Undefined, event
+                )
 
-            return
+            else:
+                # The entire list has changed. We reset the cache and fire a
+                # normal trait changed event.
+                self._update_cache(obj, trait_name)
```

Comment:
This causes the listener to hold a reference to the ExtensionPoint trait type instance. On its own, this is not a big deal, but then the trait type also tries to hold a reference to the listener to counter the weakref held in the extension registry (see the bottom of this `connect` function), so that results in a cycle reference.